### PR TITLE
Added button to renew ride

### PR DESCRIPTION
**Pull request by jensj12**
This PR adds a button in the maintenance window that renews the ride (https://openrct2.org/forums/topic/719-ride-refurbishment/). I will try to figure out how to calculate the costs properly and update it once implemented. It works now (but for free) and the button should only show up when debugging tools are enabled. Any tips are welcome.

### CODE REVIEW

File: src/windows/ride.c [hunk 1]
```diff
@@ -3396,6 +3398,9 @@ static void window_ride_maintenance_mouseup(rct_window *w, int widgetIndex)
 	case WIDX_LOCATE_MECHANIC:
 		window_ride_locate_mechanic(w);
 		break;
+	case WIDX_RENEW_RIDE:
+		game_do_command(0,1,0,w->number,GAME_COMMAND_RENEW_RIDE,0,0);
```
Comment (Gymnasiast) [inline src/windows/ride.c]: Needs spaces after commas.

**Comment by Gymnasiast** [comment]
Has duplicate code, as 'Renew rides' (in the Cheats window) does exactly the same, but for all rides at once. Could you refactor it so the existing cheat calls your function for every ride? Also, it has to be checked if this `unreliability_factor` has to be reset as well.

**Comment by duncanspumpkin** [comment]
I don't see why this has to be a new game command. There is already a ride set param game command

**Comment by Gymnasiast** [comment]
@duncanspumpkin Under which class is the 'Force breakdown' command? This feature fall in the same category, I'd say.

**Comment by Broxzier** [comment]
I like this feature, it makes rides. It would be nice if this would also cost money, otherwise it's pretty much an alternative way of using the [renew rides cheat](https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/cheats.c#L141-L154).

I also don't think this feature belongs to debugging, but is something regular users can use as well. Only thing is that it needs to calculate the cost to really make a difference between the cheat and this button.

The button should be disabled for when a ride is not old at all, and has a reliability of 100%. this prevents the player accidentally clicking the button twice, and losing a huge sum of cash.

File: data/language/en-GB.txt [hunk 2]
```diff
@@ -4114,6 +4114,7 @@ STR_5805    :{SMALLFONT}{BLACK}If checked, your server will be added to the{NEWL
 STR_5807    :{WINDOW_COLOUR_2}Number of rides: {BLACK}{COMMA16}
 STR_5808    :{WINDOW_COLOUR_2}Number of shops and stalls: {BLACK}{COMMA16}
 STR_5809    :{WINDOW_COLOUR_2}Number of information kiosks and other facilities: {BLACK}{COMMA16}
+STR_5810    :{SMALLFONT}{BLACK}Makes the ride brand new.{NEWLINE}Resets ride age and reliability
```
Comment (IntelOrca) [inline data/language/en-GB.txt]: Given you have implemented this like a gameplay feature, I find this far too technical for a user tooltip. It should say something like, refurbishes the ride.

**Comment by IntelOrca** [comment]
I agree with @Broxzier, this is just a cheat at the moment. It should cost money to renew a ride, at least the amount that it costs to build it again.

File: data/language/en-GB.txt (hunk 2 above)
Comment (Gymnasiast) [inline data/language/en-GB.txt]: He has implemented it as a debugging tools feature. Though rewording might be needed if we were to follow Broxzier's idea.

**Comment by jensj12** [comment]
It is meant to cost money, just couldn't code that yet. Once that's done, I'll make it available without debugging tools.
Didn't know about the set_ride_param game command yet, will fix that, as well as the refactor the renew rides cheat.

**Comment by Broxzier** [comment]
I can see this being abused for merged rides, if you were to only take the cost of the ride's track into account. And if you were to follow the track and remove and place each track piece again, then what about rides with multiple circuits?

Example

**Comment by jensj12** [comment]
@Broxzier Merged rides are not possible in 'normal' gameplay. If cheat users want to play fair, they can refurbish all rides in the circuit.
`unreliabily_factor` seems to be a static value that is multiplied by a value ranging from 1 to 3 (depending on the age) when it's used (https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/ride/ride.c#L2196). Resetting ride age and reliability is enough. I've not checked if the popularity/statisfaction needs to be reset.
I'm still lost at what game command to use, none of the existing ones (in ride.c) make sense to me.

**Comment by IntelOrca** [comment]
@Broxzier I think that will have to remain a technical restriction until we create our own save format which allows us to use different sprites for track elements associated with a particular ride ID.

**Comment by jensj12** [comment]
Which game command should I use? 'Force breakdown' doesn't use any, _set_ride_param doesn't exist and _set_ride_status/setting are for open/close and the ride mode / settings tab. Should I create game_command_set_ride_param and make 'force breakdown' use that one too?

**Comment by duncanspumpkin** [comment]
I was meaning set ride setting but if you think that wouldn't make sense keep as is

**Comment by jensj12** [comment]
The cost calculation is based on the sell price. The sell price of a ride is 70% and refurb price is half of that, so it nicely adds up to 105% (and thus sell + rebuild is slightly cheaper).

**Comment by Broxzier** [comment]
There are quite some merge commits in this PR now, make sure to rebase this on develop.

File: data/language/en-GB.txt [hunk 3]
```diff
@@ -4117,6 +4117,8 @@ STR_5808    :{WINDOW_COLOUR_2}Number of shops and stalls: {BLACK}{COMMA16}
 STR_5809    :{WINDOW_COLOUR_2}Number of information kiosks and other facilities: {BLACK}{COMMA16}
 STR_5810    :Disable train length limit
 STR_5811    :{SMALLFONT}{BLACK}If checked, you can have up to{NEWLINE}255 cars per train
+STR_5812    :{SMALLFONT}{BLACK}Refurbishes the ride
```
Comment (Nubbie) [inline data/language/en-GB.txt]: The tip still seems a bit unclear as it's almost the exact thing as the other line
Maybe something like: 
```
STR_5812    : Rebuilds the ride to make the reliability go back to 100% for a price
```

Comment (Nubbie) [inline data/language/en-GB.txt]: Might be a bit unclear so maybe can add:
```
STR_5812    :{SMALLFONT}{BLACK}Refurbishes the ride{NEWLINE}which sets the reliability back to 100% for a price
```

Comment (IntelOrca) [inline data/language/en-GB.txt]: That's too technical, its better to describe the overall outcome in doing it - like Refurbishes the ride to help attract more guests.

Comment (Nubbie) [inline data/language/en-GB.txt]: Wouldn't it be better to explain what it does?

Comment (jensj12) [inline data/language/en-GB.txt]: I'd like to include that it costs something, something like `Refurbishes the ride,{NEWLINE}makes it like new at a cost`

Comment (IntelOrca) [inline data/language/en-GB.txt]: It might as well tell you how much its going to cost if it is to be helpful in any way.

Comment (Broxzier) [inline data/language/en-GB.txt]: Or display it in a confirmation prompt, and if there is one then the button does not have to be disabled either.

**Comment by jensj12** [comment]
The button now makes use the demolish ride game command instead of it's own (so it also includes some proper checks). I couldn't rename it yet because of some external links I can't edit.
The game pops up a confirmation dialog (stolen from the demolish ride prompt) which shows how much it costs. Only one demolish/refurbish ride prompt can be opened at the same time, it switches nicely if another refurbish/demolish button is clicked.
Unless someone else finds something to improve, I think it's ready to be merged.

**Comment by jensj12** [comment]
Even though the button works as far as I've tested, I'm not sure about the no money scenario's. Using this would be cheating as players can keep using it for free to prevent a lot of breakdowns. Three possible solutions (it should always work when the goal is 'Have fun!' or the goal is already reached):
1. Add a warning that using the function would be cheating
2. Only allow it once the ride is at least _n_ years old, show the 'ride doesn't need refurbishing' error if not.
3. Disable it altogether (like I've done with mazes)

**Comment by IntelOrca** [comment]
I'd go with option 2

**Comment by Gymnasiast** [comment]
You missed some ride types. Crooked Houses never get broken either, and there might be more.

**Comment by IntelOrca** [comment]
So are we saying that this is only to improve the reliability of the ride and not affect whether guests choose to go on it or not?

**Comment by Broxzier** [comment]
`ride->build_date = gDateMonthsElapsed;`

It resets the ride age, which affects the likeliness of people wanting to ride it.

**Comment by jensj12** [comment]
The problem with mazes is that I can't calculate the cost of them without modifying the ride.
Currently it allows the refurbishment of rides as soon as they have been opened, as it can be helpful to reset the sell price to 100% when (re)building some parts of the coaster.

**Comment by IntelOrca** [comment]
Now that I think about it more, I don't think it should be this easy to attract guests again to a ride. I think the player *should* have to build a new ride and kind of like IRL, you sell old rides and build new better ones (probably because you have more money and the money you sold the old ride for). With a better peep AI which we will do at some point we can help distribute the guests around the rides more.

So I think renew ride should only increase reliability to reduce breakdown, and its only available if the ride is maybe 1 or 2 years old... or if reliability is lower than 50% or something.

**Comment by Wuzzy2** [comment]
I support this idea, as long it is really just a more accessible alternative to manually demolishing the ride, then building the exact same track on the spot.
When you play a park for a very long time, your old custom-built roller coasters can seriously bring down your park value, and it is unpleasant to renew them manually (even if I could totally afford it), so usually I decide to keep them. So yeah, this is a good improvement over RCT2.

Does it really cost the same as if you would manually demolish the track (get cash) and then manually build the same track again at the exact same spot (cost cash)?
This is important, I think, because any cost difference would probably divert too much from RCT2 gameplay.
Maybe there are other RCT2 aspects which need to be considered when a *new* ride is created this way (e.g. resetting of crash statistics).

I have a small suggestion: Display the cost for renewing the ride (build cost of new ride minus sell earnings of old ride) because renewing an entire rollercoaster can sometimes get *very* expensive.

**Comment by mrtnptrs** [comment]
I think this is a really nice addition for now as cheats, but.... is this ready to be merged?

**Comment by Broxzier** [comment]
@mrtnptrs: This is _not_ a cheat, it is a new feature. The only 'advantage' it gives the player is that it does not have to be rebuilt manually, and that no scenery will be lost from that progress.

**Comment by mrtnptrs** [comment]
@Broxzier I know, sorry, was still reading through the comments. But I think that it should be possible to renew the ride, even when it is less then 2 years old.

**Comment by jensj12** [comment]
I've checked multiplayer this time and it works properly.

**Comment by Broxzier** [comment]
How does this handle the costs for underground track pieces?

**Comment by GrahamRCT** [comment]
If I may make so bold as to comment on this PR without being a dev :innocent: 

Why does this cost 35%? Refurbishing is essentially rebuilding the exact same ride (but ofc allowing custom track designs to be rebuilt, which is great).

Since the reliability goes up (saves mechanics and down-time), as well as the price you can charge for your rides and park entrance, I think it's only fair you pay 100% of the ride's cost to refurbish it, no?

**Comment by jensj12** [comment]
Destroying and rebuilding a ride only costs 30% of the original price. Charging 5% to use a button seems like a fair price to me (and is easy to calculate). The 1 or 2 year limits are there for two reasons: prevent players from wasting cash on a ride they've just refurbished and to prevent the 'renew rides' cheat to be included in normal no cash scenarios.

Underground track pieces also cost 35% to refurb if I understand the game mechanics correctly.

The reason this branch is stalled is that my local code base is slightly messed up and I have zero spare time to fix it ATM. I'm not particularly happy with the price calculation process as it copies some code and I doubt if it's worth fixing the conflicts instead of starting over.

**Comment by GrahamRCT** [comment]
> Destroying and rebuilding a ride only costs 30% of the original price

Oh, right, you always get refunded 70% for aged rides!

> The 1 or 2 year limits are there for two reasons: prevent players from wasting cash on a ride they've just refurbished and to prevent the 'renew rides' cheat to be included in normal no cash scenarios.

The first reason relies on there being no reason to refurbish a ride before 1 or 2 years. But the first age penalty sets in after 4 months, so theoretically you might want to refurbish the ride at that point. 

As for the second reason: I see refurbishment as a shortcut for destroying and rebuilding a ride, which is always possible for non-custom rides. So even with the 1-2 year limit, I could replace, say, my Pirate Ship after 4 months (just without the refurbish button). Why not my custom Car Ride that goes underground?

**Comment by AaronVanGeffen** [comment]
@jensj12 With #7164 in, will you be working on this again? You can rebase this PR, or open a new one. Personally, I prefer the former.

**Comment by Gymnasiast** [comment]
We might finally see the oldest open PR merged :P

**Comment by jensj12** [comment]
I first want to have the code for obtaining the sell price sorted out and have some more thoughts about hijacking the demolish ride prompt code, but I will finish this someday.
Note to self: also update lifecycle flags

**Comment by jensj12** [comment]
I've completely refurbished this PR to fit the latest code base. Apart from some minor cleanups, it should be done and function almost the same as before. The requirement for refurbishing is now that the ride must be closed and empty (same as for rebuilding the ride, except that cars may stay on track) and the crash status will also be reset.
If a file has to be created or renamed, I kindly request some assistance, as VS won't let me. I'll happily move or rename code if needed.

**Comment by jensj12** [comment]
Tested multiplayer and multiple rides including a freefall tower, and they all work as expected.

File: src/openrct2-ui/windows/DemolishRidePrompt.cpp [hunk 4]
```diff
@@ -26,7 +26,7 @@
 #define WW 200
 #define WH 100
 
-money32 gDemolishRideCost;
+money32 _demolishRideCost;
```
Comment (AaronVanGeffen) [inline src/openrct2-ui/windows/DemolishRidePrompt.cpp]: This variable is only needed in this window, right? Rather than exporting it, just make it static.

**Comment by AaronVanGeffen** [comment]
Unfortunately, we didn't get around to reviewing your PR in time... It has some merge conflicts again. I tried pushing a rebased version, but it looks like the right permissions have not been granted on your repository.

Could you rebase your PR once more? Sorry for the wait.

**Comment by Broxzier** [comment]
I've just rebased and tested this PR with multiple parks. I've tested it with and without money, different rides, and even ones with merged parts and arbitary ride types. Only merged rides were very cheap, but as stated before those can not be achieved through normal gameplay. Everything else works nicely, and the prices are reasonable, so I reckon this can (finally) get merged.

**Comment by Gymnasiast** [comment]
Don't merge it, want to review it.

**Comment by AaronVanGeffen** [comment]
In general, I'd like to echo @Broxzier in that it seems to work quite well, and well-polished at that. Nice work.

I'm not too sure about how this is implemented within the `RideDemolishAction` GameAction, though. It would be better for the refurbishment to be an action of its own. However, I don't want to block the merger because of this.

File: src/openrct2/localisation/StringIds.h [hunk 5]
```diff
@@ -3910,6 +3910,16 @@ enum {
     STR_VIEW_CLIPPING_SELECT_AREA = 6241,
     STR_VIEW_CLIPPING_CLEAR_SELECTION = 6242,
 
+	STR_REFURBISH_RIDE_TIP = 6243,
```
Comment (Gymnasiast) [inline src/openrct2/localisation/StringIds.h]: Use spaces, not tabs.

**Comment by AaronVanGeffen** [comment]
Did another test on multiplayer. State is synced properly, but the confirmation window is not closed automatically when clicking confirm, even though the refurbishment takes place. This does not happen in single player mode.

File: src/openrct2/ride/Ride.cpp [hunk 6]
```diff
@@ -5476,114 +5476,77 @@ sint32 ride_is_valid_for_open(sint32 rideIndex, sint32 goingToBeOpen, sint32 isA
...
-        sint32 x = it.x * 32, y = it.y * 32;
-        sint32 z = it.element->base_height * 8;
+    // Find the start in case it is not a complete circuit
+    ride_get_start_of_track(&trackElement);
 
-        uint8 rotation = tile_element_get_direction(it.element);
-        uint8 type = track_element_get_type(it.element);
+    uint8 direction = tile_element_get_direction(trackElement.element);
 
-        if (type != TRACK_ELEM_INVERTED_90_DEG_UP_TO_FLAT_QUARTER_LOOP){
```
Comment (AaronVanGeffen) [inline src/openrct2/ride/Ride.cpp]: This check appears to have been lost, perhaps in a rebase at some point?

File: src/openrct2/ride/Ride.cpp [hunk 7]
```diff
@@ -5476,114 +5476,77 @@ sint32 ride_is_valid_for_open(sint32 rideIndex, sint32 goingToBeOpen, sint32 isA
...
-            }
-            tile_element_iterator_restart_for_tile(&it);
-            continue;
-        }
+    // Used in the following loop to know when we have
+    // completed all of the elements and are back at the
+    // start.
+    rct_tile_element *initial_map = trackElement.element;
 
-        // Using GAME_COMMAND_FLAG_2 for below commands as a HACK to stop fences from being removed
```
Comment (AaronVanGeffen) [inline src/openrct2/ride/Ride.cpp]: These checks appears to have been lost, perhaps in a rebase at some point?

File: src/openrct2/ride/Ride.cpp (hunk 6 above)
Comment (jensj12) [reply src/openrct2/ride/Ride.cpp]: This check is still there, but [here](https://github.com/OpenRCT2/OpenRCT2/blob/77b09b3daffb7c0ba972bc5b71e1b2af5284b1ef/src/openrct2/actions/RideDemolishAction.hpp#L271). The old ride_get_refund_price function was no longer used (and had a wrong name, should have been named DemolishTracks), so I completely rewrote that function to do what it says and used it.
TLDR: the removed code in ride_get_refund_price has already been moved.

**Comment by AaronVanGeffen** [comment]
@jensj12 It appears you're right. I was afraid that refunding mazes would be broken by the code removal, but it looks like demolishing them wasn't refunding any money without your code, either.

That removes my main objection to merging this. :)
